### PR TITLE
Validate chunked loss chunking at config init

### PR DESCRIPTION
**Pull request by fallintoplace**
## Summary

Reject `ChunkedCELoss` configs whose local sequence length would not divide evenly by `num_chunks`, using trainer config-time validation instead of a runtime tensor-shape check.

## Details

- Add `ChunkedCELoss.validate_sequence_chunking()` to validate global sequence length, sequence-dimension shard degree, and chunk count together.
- Call the validation from `Trainer.Config.__post_init__` for `ChunkedCELoss` configs. At the chunked loss boundary, hidden states have passed through the root decoder norm: TP sequence shards are gathered before `lm_head`, while CP remains sharded on the sequence dimension, so the loss-local shard degree is `context_parallel_degree`.
- Keep `GradAccumulator` defensive by validating equal local chunk sizes if it is used directly.
- Update tests to cover uneven local chunks, uneven sequence sharding, and valid sharded chunking.

## Validation

- `python -m pytest -q tests/unit_tests/test_loss.py`
- `SKIP=pyrefly-check pre-commit run --files torchtitan/components/loss.py torchtitan/trainer.py tests/unit_tests/test_loss.py`
- `pre-commit run --files torchtitan/components/loss.py torchtitan/trainer.py tests/unit_tests/test_loss.py` currently fails only in the project-wide Pyrefly hook on unrelated existing errors in `torchtitan/hf_datasets/multimodal/mm_collator.py`, `torchtitan/models/common/moe.py`, `torchtitan/models/flux/model/autoencoder.py`, `torchtitan/models/qwen3_5/vision_encoder.py`, and `torchtitan/distributed/utils.py`; all other hooks pass.

### CODE REVIEW

File: torchtitan/components/loss.py [hunk 1]
```diff
@@ -411,13 +412,14 @@ def add(self, chunk_grad: torch.Tensor) -> None:
             chunk_grad = chunk_grad.to(self._buffer.dtype)
 
         chunk_seq_len = chunk_grad.shape[self.seq_dim]
-        start = self._next_idx * chunk_seq_len
+        start = self._seq_offset
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: Thanks!

The fix sounds correct. But I wonder if we should support chunking & accumulating an already short sequence. How about just raise in `ChunkedCELoss` when `num_chunks` cannot evenly divide seq_len?

File: torchtitan/components/loss.py [hunk 2]
```diff
@@ -530,6 +547,23 @@ def __call__(
...
+            if isinstance(hidden_states, DTensor)
+            else hidden_states.shape[1]
+        )
+        if local_seq_len % num_chunks != 0:
+            raise ValueError(
+                "ChunkedCELoss requires the local sequence length to be evenly "
+                f"divisible by num_chunks, got local sequence length {local_seq_len} "
+                f"and num_chunks {num_chunks}."
+            )
+
```
Comment (pianpwk) [inline torchtitan/components/loss.py]: I think it's possible (and likely better) to unify as an init-time rather than runtime check, based on global sequence length and seq-dim sharding degree.

IIRC, both CP and TP (if SP on) will shard the activation seq-dim, so `global_seq_len / (cp_degree * tp_degree if enable_sp else 1)` gives you local_seq_len. Then if uneven sharding is allowed (I don't recall, will need to check), the last rank will also have a remainder_seq_len. Then you can verify by testing divisibility by num_chunks for those values.

File: torchtitan/components/loss.py [hunk 3]
```diff
@@ -484,14 +501,53 @@ class ChunkedCELoss(BaseLoss):
...
+        if num_chunks <= 0:
+            raise ValueError(
+                f"ChunkedCELoss num_chunks must be positive, got {num_chunks}"
+            )
+        if seq_len % seq_shard_degree != 0:
+            raise ValueError(
+                "ChunkedCELoss requires the global sequence length to be evenly "
+                f"divisible by the sequence shard degree, got sequence length "
+                f"{seq_len} and sequence shard degree {seq_shard_degree}."
+            )
```
Comment (pianpwk) [inline torchtitan/components/loss.py:530]: i'm actually not sure if this check is too strict, local seqlen evenness is required, but global / sp_degree is nice-to-have, but maybe not necessary ?

File: torchtitan/components/loss.py [hunk 4]
```diff
@@ -484,14 +501,53 @@ class ChunkedCELoss(BaseLoss):
...
+
+        local_seq_len = seq_len // seq_shard_degree
+        if local_seq_len % num_chunks != 0:
+            raise ValueError(
+                "ChunkedCELoss requires the local sequence length to be evenly "
+                f"divisible by num_chunks, got global sequence length {seq_len}, "
+                f"sequence shard degree {seq_shard_degree}, local sequence length "
+                f"{local_seq_len}, and num_chunks {num_chunks}."
+            )
```
Comment (pianpwk) [inline torchtitan/components/loss.py:539]: the rest looks good to me, will wait for @tianyu-l to approve init-time validation

File: torchtitan/components/loss.py [hunk 5]
```diff
@@ -375,6 +375,17 @@ def __init__(
         else:
             local = reference
 
+        if num_chunks <= 0:
+            raise ValueError(f"num_chunks must be positive, got {num_chunks}")
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:379]: don't need this

File: torchtitan/components/loss.py [hunk 6]
```diff
@@ -375,6 +375,17 @@ def __init__(
...
+
+        seq_len = local.shape[seq_dim]
+        if seq_len % num_chunks != 0:
+            raise ValueError(
+                "GradAccumulator requires the sequence length to be evenly "
+                f"divisible by num_chunks, got sequence length {seq_len} "
+                f"and num_chunks {num_chunks}."
+            )
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:387]: this has been tested in trainer config post_init, maybe just `assert` here?

File: torchtitan/components/loss.py [hunk 7]
```diff
@@ -411,6 +422,12 @@ def add(self, chunk_grad: torch.Tensor) -> None:
             chunk_grad = chunk_grad.to(self._buffer.dtype)
 
         chunk_seq_len = chunk_grad.shape[self.seq_dim]
+        if chunk_seq_len != self._chunk_seq_len:
+            raise ValueError(
+                "GradAccumulator expected chunk sequence length "
+                f"{self._chunk_seq_len}, got {chunk_seq_len}."
+            )
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:429]: same here, maybe `assert` is enough, as I don't think user could hit it anyways

File: torchtitan/components/loss.py [hunk 8]
```diff
@@ -484,14 +501,53 @@ class ChunkedCELoss(BaseLoss):
...
     def __init__(
         self,
         config: Config,
         *,
         compile_config: CompileConfig | None = None,
     ):
+        if config.num_chunks <= 0:
+            raise ValueError(
+                f"ChunkedCELoss num_chunks must be positive, got {config.num_chunks}"
+            )
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:550]: move this to post_init of ChunkedLoss.Config

File: torchtitan/components/loss.py [hunk 9]
```diff
@@ -484,14 +501,53 @@ class ChunkedCELoss(BaseLoss):
     @dataclass(kw_only=True, slots=True)
     class Config(BaseLoss.Config):
         num_chunks: int = 8
-        """Number of chunks to split the sequence into."""
+        """Number of chunks to split the local sequence into."""
+
+    @staticmethod
+    def validate_sequence_chunking(
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:507]: I think this can be an instance method of `ChunkedCELoss.Config`

File: torchtitan/trainer.py [hunk 10]
```diff
@@ -157,6 +157,16 @@ def __post_init__(self):
...
 
+            if isinstance(self.loss, ChunkedCELoss.Config):
+                # ChunkedCELoss sees hidden states after the root decoder norm.
+                # That norm gathers TP sequence-parallel shards before lm_head,
+                # while CP keeps the local sequence dimension sharded.
+                ChunkedCELoss.validate_sequence_chunking(
+                    seq_len=self.training.seq_len,
+                    seq_shard_degree=self.parallelism.context_parallel_degree,
+                    num_chunks=self.loss.num_chunks,
+                )
```
Comment (tianyu-l) [inline torchtitan/trainer.py:168]: so that you could do `self.loss.validate_...(seq_len=, cp_degree=)`
